A508/509/512 gpucc and speedbins rework

minlexx opened the pull request:
* Fix reading the speedbins correctly.
* Use `ADRENO_SPEEDBINS` to specify fuses and their corresponding bit shift values in `a5xx_catalog.c`
* Move `a6xx_set_supported_hw()` to common code, renaming it to `adreno_set_supported_hw()`
* Use new common funcs to setup opp-supp-hw on a5xx
* Update sdm630/636/660 SoC device trees to use the correct supp-hw values, uncomment all freqs in opp-table
* Stop doing custom opp-table overrides in individual device `.dts`
* Fix gpucc-sdm660 driver to actually allow switching frequencies, using new dedicated clk_gfx3d_src_ops. Add frequency tables, same way as it's done in downstream

TBD

P.S.
~~* sony-xperia-nile fixes are irrelevant here, should go into separate branch~~
* the last commit "DEBUG drop-me: opp/of: debug prints" should go away
* the revert commit of "gpu go 700mhz" and the revert target commit should be dropped before release

minlexx committed:
Revert "gpu go 700mhz"

This reverts commit 2f4ea689ded3237cca472dac871cb4aef635b733.

minlexx committed:
DEBUG drop-me: opp/of: debug prints

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

minlexx committed:
drm/msm/adreno: Move speedbin functions from a6xx to common file

In order for a5xx code to be able to use the same method of
configuring speedbins and opp-supported-hw, move fuse_to_supp_hw()
and a6xx_set_supported_hw() (renamed to adreno_set_supported_hw) to
common code into adreno_gpu.c.

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

minlexx committed:
drm/msm/a5xx: Specify speedbins for A508/509/512

Specify speedbins for Adreno 508, 509 and 512 GPUs found in
SDM630/636/660 SoCs and their respective bits used for bitmasks
in opp-supported-hw setting, using ADRENO_SPEEDBINS macro, same
way as a6xx_catalog does it.

This will allow us to reuse common code for a5xx/a6xx to configure
opp tables and its supported-hw values.

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

minlexx committed:
drm/msm/a5xx: Use common method to setup opp-table supported hw

For Adreno 508/509/512 use the newly available common method to
configure opp-tables supported hw value, shared with a6xx.

Keep the old way (check_speed_bin) around for plaforms that have
not yet been ported device trees and definion of speed bins in
a5xx_catalog. This way only SDM630/636/660 are affected, to
avoid potential regressions on e.g. MSM8996/8998.

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

barni2000 commented on drivers/clk/qcom/clk-rcg2.c:
```diff
@@ -1318,6 +1318,114 @@ const struct clk_ops clk_gfx3d_ops = {
+	parent_req.best_parent_rate = clk_hw_get_rate(parent_req.best_parent_hw);
+
+
+	//pr_info("pre_div: %d freq: %lu parent_req->rate: %lu\n", f->pre_div, f->freq, parent_req.rate);
```
It should be removed later

minlexx commented on drivers/gpu/drm/msm/adreno/adreno_gpu.c:
```diff
@@ -1085,6 +1085,55 @@ int adreno_read_speedbin(struct device *dev, u32 *speedbin)
+	supp_hw = fuse_to_supp_hw(info, speedbin);
+
+	dev_info(dev, "GPU speedbin fuse %d (0x%x), mapped to opp-supp-hw 0x%x\n",
+		 speedbin, speedbin, supp_hw);
```
This debug print I added myself, this was not originally present in a6xx_gpu before moving, so perhaps it should be removed too

minlexx replied:
I'd better extract this debug print to extra commit that can be easily dropped during some rebase

minlexx committed:
fixup! arm64: dts: qcom: Add initial support for asus-x00td

minlexx committed:
fixup! arm64: dts: qcom: add device-tree for Redmi Note 6 Pro (tulip)

minlexx committed:
fixup! arm64: dts: qcom: add device-tree for Redmi Note 5 Pro (whyred)

minlexx committed:
arm64: dts: qcom: sdm630: Fix GPU speedbin definition

Currently GPU speedbin efuse is not read correctly. On my device
(sdm660-xiaomi-laevnder) the resulting value in Adreno driver is
0x7. There is no such speedbin in [1]. It should read 0x87 (135)
which corresponds to dowsntream's qcom,gpu-pwrlevels-3 with 647
MHz max, which is further confirmed by testing on the device
(by uncommenting and hardcoding opp-table frequencies).

When you compare :
 0x7  (0x00000111)
 0x87 (0x10000111)

it seems that the higher bit is lost somewhere. Also when you look
at the list of speedbins on each SoCs:

 * SDM636/660: 157 (0x9d), 146 (0x92), 135 (0x87), 122 (0x7a),
                90 (0x5a),  78 (0x4e)
 * SDM630: 162 (0xa2), 146 (0x92), 135 (0x87)

it becomes clear that 7 bits is not enough to hold values above 127.
We need 8 bits. But changing only bitmask width is not enough to get
the correct value. Reg size needs to increased to 2 bytes, because
the bits that we need to read are spread between 2 bytes: 3 highest
bits of 0x41a2 and 5 lower bits of 0x41a3.

With this change we can finally get the correct GPU speedbin value.

Validity of speedbin also can be confirmed if you boot downstream
(Android or recovery) kernel and do

 cat /sys/kernel/gpu/gpu_max_clock

On e.g. xiaomi-lavender it will show: 647, which corresponds to
maximum frequency for bin 135.

Tested by me on 4 different devices:
* sdm660-xiaomi-lavender: bin 135
* sdm660-xiaomi-clover-plus: bin 135
* sdm660-xiaomi-jason: bin 135
* sdm636-xiaomi-whyred: bin 90
* sdm630-sony-nile-pioneer: bin 146.

Other people in sdm660-mainline comunity have tested on their
devices too, for statistics I'll show the results:

 * sda660-xiaomi-clover (NOT PLUS): bin 135
 * sdm660-xiaomi-jasmine_sprout: bin 135
 * sdm660-xiaomi-lavender (variant with other panel): still bin 135
 * sdm636-asus-x00td: bin 90
 * sdm636-motorola-beckham: bin 90
 * sdm630-motorola-payton: bin 146.

So far, with all devices we were able to test, all SDM660 devices
have bin 135 with 647 MHz max, all SDM636 devices have bin 90 with
430 MHz max, and all SDM630 devices have bin 146 with 700 MHz max.

References used:
[1] https://git.codelinaro.org/clo/la/kernel/msm-4.4/-/blob/LA.UM.7.2.c27-07400-sdm660.0/arch/arm/boot/dts/qcom/sdm660-gpu.dtsi
[2] https://git.codelinaro.org/clo/la/kernel/msm-4.4/-/blob/LA.UM.7.2.c27-07400-sdm660.0/arch/arm/boot/dts/qcom/sdm630-gpu.dtsi

P.S. Close relative sm6115.dtsi is also using reg size=2 and bitmask
width 8 for it's speedbin fuse definition, so we''l not be the first.

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

minlexx committed:
arm64: dts: qcom: sdm630/660: Fix GPU opp-tables

After gpucc-sdm660 rework, make frequencies list in opp-tables
match the ones in GPU clock controller driver for each of SDM630
and SDM660 platforms. They should match to avoid error messages
in dmesg like:

  dev_pm_opp_set_rate: failed to find OPP for freq 500000000 (-34)

Previous authors left highest frequencies commented due to some
issues with interconnect framework and missing support for the
GPU regulator; keep them commented, uncomment all others, add
19.2 MHz one, specify correct opp-supported-hw values for each
one of them.

Since list of frequencies supported by GPUCC is different between
SDM630 and SDM660, and there is also difference in frequencies
supported on bin 135:

 * SDM660, bin 135: 647000000, 588000000, 465000000, 370000000,
                    266000000, 160000000, 19200000
 * SDM630, bin 135: 647000000, 588000000, 465000000, 370000000,
                    240000000, 160000000, 19200000

(266 MHz vs 240 MHz), so we have to delete the opp-table inherited
from sdm630.dtsi and write new one for sdm660/636 from scratch.

P.S. Maybe we can uncomment all high OPPs, because they are not
available due to speedbin values?

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

barni2000 commented on arch/arm64/boot/dts/qcom/sdm660.dtsi:
```diff
@@ -13,61 +13,82 @@
 		compatible = "operating-points-v2";
 
 		/*
 		 * 775MHz is only available on the highest speed bin
```
```suggestion
		 * 750MHz is only available on the highest speed bin
```

barni2000 committed:
clk: qcom: rcg2: Add new gfx3d_src ops for SDM630/660 GPUCC

Add new specific version of RCG2 ops (clk_gfx3d_src_ops) that is
designed to support GPU clock controller driver for SDM630/660 SoCs.
Already existing clk_gfx3d_ops doesn't support handling frequency
tables for gfx3d_src_clk, which is fine for Adreno 530/540 with GMU,
but is required for lower-end A508/509/512 without GMU.

Co-developed-by: Barnabás Czémán <barnabas.czeman@mainlining.org>
Signed-off-by: Barnabás Czémán <barnabas.czeman@mainlining.org>
Co-developed-by: Alexey Minnekhanov <alexeymin@postmarketos.org>
Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

barni2000 committed:
clk: qcom: gpucc-sdm660: Add support for real frequency switching

For platforms without GMU frequency for gfx3d clock needs to be
managed by GPU devfreq withing adreno driver. Currently there is
a problem: there is no frequency table specified for gfx3d_clk_src
and gfx3d clock is running on some "random" 500 MHz frequency, as
can be seen on SDM630/660 devices:

 $ grep -i -E '(gpu|gfx)' /sys/kernel/debug/clk/clk_summary
 $ cat /sys/kernel/debug/devfreq/devfreq_summary

Will output: (extra columns removed to fit)

 gpu_pll1_pll_out_main    740000000
 gpu_pll0_pll_out_main   1000000000
    gfx3d_clk_src         500000000
       gpucc_gfx3d_clk    500000000

 dev          cur_freq_Hz  min_freq_Hz  max_freq_Hz
 --------------------------------------------------
 5000000.gpu  700000000    700000000    700000000

And additionally, the following errors can be seen in dmesg:

 devfreq 5000000.gpu: Couldn't update frequency transition information.

As you can see, gf3d_clk is running on some weird 500 Mhz
frequency, which is not even present in GPU opp-tables, with
parent gpu_pll on 1 GHz. And additionally, devfreq thinks it's
running on 700 MHz (output is from SDM630 device).

Lucky SDM660 can function this way, since most speedbins allow
frequencies up to 647 MHz, and on SDM630 700 MHz max is commonly
seen. Can function, but constantly only on 500 MHz. Unlucky SDM636
devices, however, can't probe GPU at all: observing either system
freeze during boot, or GPU hangs. Sometimes it can randomly probe
with gx3d_clk parented to gpucc_cxo running at 19.2 MHz and producing
2-5 FPS in GUI mobile shells like phosh. Which currently puts all
SDM636 devices into the category of "barely functioning bricks".

To solve all this, we need:
 * add freq table to gfx3d_clk_src;
 * switch to new clk_gfx3d_src_ops to support all this.

With this change, SDM636 phones turn from unusable bricks to
almost functional linux mobile devices (given the other missing
parts of the platform support); and SDM630/660 phones receive
the ability to run sometimes at slightly higher speeds, giving
slightly more FPS in SuperTuxKart and more points in glmark2.
Additionally on all of the SDM630/636/660 gpucc_gfx3d_clk will
be able to run on the lowest speeds (down to "idle" 19.2 MHz)
which I believe should help power savings.

Also, there will be no discrepancy between GPU devfreq and real
clock frequency, and no errors about missing frequencies from
opp tables.

Co-developed-by: Barnabás Czémán <barnabas.czeman@mainlining.org>
Signed-off-by: Barnabás Czémán <barnabas.czeman@mainlining.org>
Co-developed-by: Alexey Minnekhanov <alexeymin@postmarketos.org>
Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

minlexx committed:
DEBUG adreno: show speedbin in dmesg

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

minlexx committed:
drm/msm/a5xx: Allow to read speedbin value from debugfs

Add "speedbin" file to msm display controller debugfs.
Knowing this value is useful for debugging.

Works like this:

 $ cat /sys/kernel/debug/dri/c901000.display-controller/speedbin
 135

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>
